From ac1e8443a250f418b6124e7b4f4ea65a03c4d02b Mon Sep 17 00:00:00 2001
From: Benjamin Lin <benjamin-jw.lin@mediatek.com>
Date: Fri, 26 Apr 2024 09:29:39 +0800
Subject: [PATCH] mac80211: mtk: add exported function for SoftMAC driver to
 get QoS map

The mapping from IP DSCP to IEEE 802.11 user priority may be customized.
Therefore, driver needs to pass the mapping to HW, so that the QoS type of traffic can be mapped in a consistent manner for both SW and HW paths.

Signed-off-by: Benjamin Lin <benjamin-jw.lin@mediatek.com>
---
 include/net/mac80211.h | 12 ++++++++++++
 net/mac80211/util.c    | 10 +++++++++-
 2 files changed, 21 insertions(+), 1 deletion(-)

--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -7489,4 +7489,16 @@ void ieee80211_set_active_links_async(st
  */
 unsigned long ieee80211_get_scanning(struct ieee80211_hw *hw);
 
+/**
+ * ieee80211_get_qos_map - get QoS mapping information.
+ *
+ * @vif: &struct ieee80211_vif pointer from the add_interface callback.
+ *
+ * Return: Pointer to the QoS mapping information.
+ *
+ * Note that the return value is an RCU-protected pointer, so rcu_read_lock()
+ * must be held when calling this function.
+ */
+struct cfg80211_qos_map *ieee80211_get_qos_map(struct ieee80211_vif *vif);
+
 #endif /* MAC80211_H */
--- a/net/mac80211/util.c
+++ b/net/mac80211/util.c
@@ -5158,3 +5158,10 @@ unsigned long ieee80211_get_scanning(str
 }
 EXPORT_SYMBOL(ieee80211_get_scanning);
 
+struct cfg80211_qos_map *ieee80211_get_qos_map(struct ieee80211_vif *vif)
+{
+	struct mac80211_qos_map *qos_map = rcu_dereference(vif_to_sdata(vif)->qos_map);
+
+	return qos_map ? &qos_map->qos_map : NULL;
+}
+EXPORT_SYMBOL(ieee80211_get_qos_map);
